cps/kni.c

```diff
@@ -294,6 +298,295 @@ kni_config(struct rte_kni *kni, struct gatekeeper_if *iface)
+			return MNL_CB_ERROR;
+		break;
+	case RTA_TABLE:
+	case RTA_DST:
+	case RTA_SRC:
+	case RTA_OIF:
+	case RTA_FLOW:
+	case RTA_PREFSRC:
+	case RTA_GATEWAY:
+	case RTA_PRIORITY:
+		if (mnl_attr_validate(attr, MNL_TYPE_U32) < 0)
+			return MNL_CB_ERROR;
+		break;
+	}
+	tb[type] = attr;
+	return MNL_CB_OK;
+}
+
+static int
+data_ipv6_attr_cb(const struct nlattr *attr, void *data)
+{
+	const struct nlattr **tb = data;
+	int type = mnl_attr_get_type(attr);
+
+	/* Skip unsupported attribute in user-space. */
+	if (mnl_attr_type_valid(attr, RTA_MAX) < 0)
+		return MNL_CB_OK;
+
+	switch(type) {
+	case RTA_MULTIPATH:
+		if (mnl_attr_validate(attr, MNL_TYPE_NESTED) < 0)
+			return MNL_CB_ERROR;
+		break;
+	case RTA_TABLE:
+	case RTA_OIF:
+	case RTA_FLOW:
+	case RTA_PRIORITY:
+		if (mnl_attr_validate(attr, MNL_TYPE_U32) < 0)
+			return MNL_CB_ERROR;
+		break;
+	case RTA_DST:
+	case RTA_SRC:
+	case RTA_PREFSRC:
+	case RTA_GATEWAY:
+		if (mnl_attr_validate2(attr, MNL_TYPE_BINARY,
+					sizeof(struct in6_addr)) < 0)
+			return MNL_CB_ERROR;
+		break;
+	}
+	tb[type] = attr;
+	return MNL_CB_OK;
+}
+
+static int
+route_cb(const struct nlmsghdr *nlh, void *arg)
+{
+	struct nlattr *tb[RTA_MAX + 1] = {};
+	struct cps_config *cps_conf = get_cps_conf();
+	struct rtmsg *rm = mnl_nlmsg_get_payload(nlh);
+	struct route_update *update = arg;
+
+	update->valid = false;
+
+	if (nlh->nlmsg_type != RTM_NEWROUTE &&
+			nlh->nlmsg_type != RTM_DELROUTE) {
+		RTE_LOG(NOTICE, GATEKEEPER,
+			"cps: unrecognized netlink message type: %u\n",
+			nlh->nlmsg_type);
+		return MNL_CB_OK;
+	}
+
+	if (cps_conf->debug)
+		RTE_LOG(INFO, GATEKEEPER, "cps update: [%s] family=%u dst_len=%u src_len=%u tos=%u table=%u protocol=%u scope=%u type=%u flags=%x\n",
+			nlh->nlmsg_type == RTM_NEWROUTE ? "NEW" : "DEL",
+			rm->rtm_family, rm->rtm_dst_len, rm->rtm_src_len,
+			rm->rtm_tos, rm->rtm_table, rm->rtm_protocol,
+			rm->rtm_protocol, rm->rtm_scope, rm->rtm_flags);
+
+	/*
+	 * TODO Which fields and attributes are needed for the LPM update?
+	 * Add them to @update here and in attr_get(), and if all of those
+	 * fields are present, set update->valid to true before returning.
+	 */
+	update->type = nlh->nlmsg_type;
+	update->family = rm->rtm_family;
+
+	switch(rm->rtm_family) {
+	case AF_INET:
+		mnl_attr_parse(nlh, sizeof(*rm), data_ipv4_attr_cb, tb);
+		attr_get(cps_conf, update, rm->rtm_family, tb);
+		break;
+	case AF_INET6:
+		mnl_attr_parse(nlh, sizeof(*rm), data_ipv6_attr_cb, tb);
+		attr_get(cps_conf, update, rm->rtm_family, tb);
+		break;
+	default:
+		RTE_LOG(NOTICE, GATEKEEPER,
+			"cps: unrecognized family in netlink event: %u\n",
+			rm->rtm_family);
+		break;
+	}
+
+	return MNL_CB_OK;
+}
+
+void
+kni_cps_route_event(struct cps_config *cps_conf)
+{
+	struct route_update updates[MAX_CPS_ROUTE_UPDATES];
+	unsigned int num_updates = 0;
+	char buf[MNL_SOCKET_BUFFER_SIZE];
+
+	do {
+		int ret = recv(mnl_socket_get_fd(cps_conf->nl), buf,
```

Comment:
Ok, but I think it's better to copy the code of `mnl_socket_recvfrom()` and add the needed flag. I'm saying it because `mnl_socket_recvfrom()` verifies some error conditions that `kni_cps_route_event()` does not. You find the code of `mnl_socket_recvfrom()` here:
http://git.netfilter.org/libmnl/tree/src/socket.c#n263

Don't forget to add a comment mentioning where to find the original code.